File: src/prod/src/managed/Api/src/System/Fabric/health/PartitionHealthStateChunkList.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Health
{
    using Interop;
    using System.Collections.Generic;

    /// <summary>
    /// Represents a list that contains <see cref="System.Fabric.Health.PartitionHealthStateChunk"/> items.
    /// </summary>
    /// <remarks>The list can be obtained through health state chunk queries. The queries may have as result more chunks that can fit a message.
    /// In this case, the list of items that fit the message is returned plus a count that shows how many total items are available.</remarks>
    public sealed class PartitionHealthStateChunkList : HealthStateChunkList<PartitionHealthStateChunk>
    {
        /// <summary>
        /// Instantiates an empty <see cref="System.Fabric.Health.PartitionHealthStateChunkList"/>.
        /// </summary>
        public PartitionHealthStateChunkList() : base()
        {
        }

        internal PartitionHealthStateChunkList(IList<PartitionHealthStateChunk> list) : base(list)
        {
        }

        internal static unsafe PartitionHealthStateChunkList CreateFromNativeList(IntPtr nativeListPtr)
        {
            var retval = new PartitionHealthStateChunkList();

            if (nativeListPtr != IntPtr.Zero)
            {
                var nativeList = (NativeTypes.FABRIC_PARTITION_HEALTH_STATE_CHUNK_LIST*)nativeListPtr;

                var nativeItemArray = (NativeTypes.FABRIC_PARTITION_HEALTH_STATE_CHUNK*)nativeList->Items;
                for (int i = 0; i < nativeList->Count; ++i)
                {
                    var nativeItem = *(nativeItemArray + i);
                    retval.Add(PartitionHealthStateChunk.FromNative(nativeItem));
                }

                retval.TotalCount = (long)nativeList->TotalCount;
            }
            
            return retval;
        }
    }
}

File: src/prod/src/ServiceModel/PartitionHealthStateChunkList.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace ServiceModel;

INITIALIZE_SIZE_ESTIMATION(PartitionHealthStateChunkList)

StringLiteral const TraceSource("PartitionHealthStateChunkList");

PartitionHealthStateChunkList::PartitionHealthStateChunkList()
    : HealthStateChunkList()
    , items_()
{
}

PartitionHealthStateChunkList::PartitionHealthStateChunkList(
    ULONG totalCount,
    std::vector<PartitionHealthStateChunk> && items)
    : HealthStateChunkList(totalCount)
    , items_(move(items))
{
}

PartitionHealthStateChunkList::~PartitionHealthStateChunkList()
{
}

ErrorCode PartitionHealthStateChunkList::ToPublicApi(
    __in Common::ScopedHeap & heap,
    __out FABRIC_PARTITION_HEALTH_STATE_CHUNK_LIST & publicPartitionHealthStateChunkList) const
{
    publicPartitionHealthStateChunkList.TotalCount = this->TotalCount;
    auto error = PublicApiHelper::ToPublicApiList<PartitionHealthStateChunk, FABRIC_PARTITION_HEALTH_STATE_CHUNK, FABRIC_PARTITION_HEALTH_STATE_CHUNK_LIST>(
        heap,
        items_,
        publicPartitionHealthStateChunkList);
    if (!error.IsSuccess()) { return error; }

    return ErrorCode::Success();
}

Common::ErrorCode PartitionHealthStateChunkList::FromPublicApi(FABRIC_PARTITION_HEALTH_STATE_CHUNK_LIST const * publicPartitionHealthStateChunkList)
{
    if (publicPartitionHealthStateChunkList == NULL)
    {
        return ErrorCode::Success();
    }

    this->TotalCount = publicPartitionHealthStateChunkList->TotalCount;
    auto error = PublicApiHelper::FromPublicApiList<PartitionHealthStateChunk, FABRIC_PARTITION_HEALTH_STATE_CHUNK_LIST>(publicPartitionHealthStateChunkList, items_);
    if (!error.IsSuccess()) { return error; }

    return ErrorCode::Success();
}
